### PR TITLE
Relax punishment for peers relaying invalid blocks and headers

### PR DESCRIPTION
**Pull request by luke-jr**
Supercedes #10512 by simply restricting punishments to our outgoing non-feeler connections, and only punishing with a disconnect, not a ban.

This is necessary to avoid banning peers that merely run old formerly-full nodes, after a softfork. We disconnect primary peers because we want compatible full nodes for that role, but allow non-full nodes to remain connected to inbound slots so they can sync the correct chain from us.

### CODE REVIEW

**Comment by ajtowns** [comment]
This seems like a reasonable approach to me; any idea why stop_node in the zapwallettxes test isn't getting a 0 exit code?

**Comment by luke-jr** [comment]
@ajtowns Not sure what you're asking? The tests are passing...

**Comment by ajtowns** [comment]
I guess whatever I saw was a transient error then (or I was just confused)...

File: src/net_processing.cpp [hunk 1]
```diff
@@ -865,10 +889,16 @@ void PeerLogicValidation::BlockChecked(const CBlock& block, const CValidationSta
     if (state.IsInvalid(nDoS)) {
         // Don't send reject message with code 0 or an internal reject code.
         if (it != mapBlockSource.end() && State(it->second.first) && state.GetRejectCode() > 0 && state.GetRejectCode() < REJECT_INTERNAL) {
+            const NodeId node_id = it->second.first;
+            CNodeState * const nodestate = State(node_id);
             CBlockReject reject = {(unsigned char)state.GetRejectCode(), state.GetRejectReason().substr(0, MAX_REJECT_MESSAGE_LENGTH), hash};
             State(it->second.first)->rejects.push_back(reject);
```
Comment (jimpo) [inline src/net_processing.cpp]: nit: Use nodestate instead of State(it->second.first)

**Comment by TheBlueMatt** [comment]
Have we considered preferring a relatively short-time-period ban instead of purely disconnecting? If the peer is somehow braindead and tries to reconnect immediately we may prefer to keep them gone for 30 minutes or an hour instead of letting them get into some loop of connect-relay-disconnect.

**Comment by sdaftuar** [comment]
It seems like this reverts #8305 with no mechanism to replace that functionality. Rationale?  This looks like a regression.

**Comment by achow101** [comment]
It was mentioned during the IRC meeting that this already did what I am planning on doing in #11446. However this doesn't seem to cover the case that I am concerned about: receiving a header for a block we have already marked as invalid. In such a case, `nDos` will be 0 and thus the peer won't be punished for relaying us the duplicate header.

File: test/functional/sendheaders.py [hunk 2]
```diff
@@ -526,21 +526,6 @@ def run_test(self):
         # Remove the first two entries (blocks[1] would connect):
         blocks = blocks[2:]
 
-        # Now try to see how many unconnecting headers we can send
```
Comment (promag) [inline test/functional/sendheaders.py]: Test passes with the above changes.

**Comment by luke-jr** [comment]
> It seems like this reverts #8305 with no mechanism to replace that functionality. 

I don't see how. The conditions this disconnects peers is a superset of the conditions #8305 did.

File: test/functional/test_framework/mininode.py [hunk 3]
```diff
@@ -79,18 +80,38 @@ def __init__(self):
...
+
+        if self.node_outgoing:
+            logger.debug('Connecting from Bitcoin Node: %s:%d' % (self.dstaddr, self.dstport))
+
+            listen_sock = socket.socket()
+            listen_sock.bind(('127.0.0.1', 0))
+            listen_sock.listen(1)
+            listen_port = listen_sock.getsockname()[1]
+            self.rpc.addnode('127.0.0.1:%u' % (listen_port,), 'onetry', False)
+            (sock, addr) = listen_sock.accept()
```
Comment (practicalswift) [inline test/functional/test_framework/mininode.py]: Use `(sock, _) = listen_sock.accept()` to clarify that the second element in the tuple is discarded intentionally

**Review by sdaftuar — CHANGES_REQUESTED** [review]
Re-NACK.  While I generally agree with the high-level goals described in the OP, it's not clear what changes here are still necessary to achieve that goal -- several related PRs have been merged in the year it's been since this was originally opened. This PR also seems to break the bugfix made in #8305, without explanation or other handling, which seems unnecessary and unrelated to the overall goal of the PR.  And the changes included in this PR seem disjointed (with comments not being updated to reflect new behavior, old nonsensical behavior left in cases that don't make sense, etc) -- presumably because the code changes here have gotten stale?  I didn't bother leaving detailed comments in all those cases because after reviewing this, it's not at all clear to me what you're trying to accomplish.

I suggest this PR be closed; if there are any remaining changes that are still relevant and worth proposing, we can evaluate fresh in a new PR.  If you think I'm mistaken, please give more detailed documentation and explanation of the reasoning behind these changes (for instance, why does the headers processing logic need to change at all?  what cases are you trying to cover, and which cases are still left out?).

Also some of the testing-related changes are separately useful/interesting and warrant their own discussion.  In particular I think supporting outbound peers that are not treated as "manual connections" for net processing purposes would be great, but I think those changes should be reviewed on their own -- I'd prefer we split those out into their own PR, rather than have them be an afterthought to support a test in a single PR.

**Comment by luke-jr** [comment]
https://github.com/bitcoin/bitcoin/pull/10593#issuecomment-339776309 ... you never explained how you think it reverts #8305

**Comment by sdaftuar** [comment]
@luke-jr The point of #8305 was to avoid giving DoS points to peers for the occasional unconnecting header.  This PR would immediately disconnect an outbound peer when it announces a header that doesn't connect.

**Comment by MarcoFalke** [comment]
`feature_block.py` fails on all travis jobs that run it

**Comment by ajtowns** [comment]
Probably should close this PR; it needs a pretty serious rebase, the RECENT_CONSENSUS_CHANGE stuff from #15141 probably warrants some more thought about the approach which might not happen until we've got a consensus change to go with it, and the current approach has a NACK...

#12674 and the "nodo-to-test connections" support for mininode are probably relevant for #14210 though.

**Comment by luke-jr** [comment]
>  it needs a pretty serious rebase

Rebase is now complete. Push needs to wait for the PR to be reopened, but @sdaftuar is right that it should probably be split up and smaller changes merged first.

> the RECENT_CONSENSUS_CHANGE stuff from #15141 probably warrants some more thought about the approach which might not happen until we've got a consensus change to go with it,

Recent consensus changes shouldn't be (at least in this context) treated any differently than other consensus rules. You're looking at it in the limited scenario of a softfork, but it needs to be broader to ensure we overwrite/reorg any attempts to hardfork without consensus too.

**Pull request closed without merging**